src/layouts.jl

```diff
@@ -0,0 +1,139 @@
+
+_as_index(x) = x
+_as_index(x::OneTo) = static(1):length(x)
+_as_index(x::StepRange) = OptionallyStaticStepRange(x)
+_as_index(x::UnitRange) = OptionallyStaticUnitRange(x)
+_as_index(x::OptionallyStaticRange) = x
+
+"""
+    StrideLayout(A)
+
+Produces an array whose elements correspond to the linear buffer position of `A`'s elements.
+"""
+struct StrideLayout{N,R,O1,S,A<:Tuple{Vararg{Any,N}}} <: AbstractArray2{Int,N}
+    rank::R
+    offset1::O1
+    strides::S
+    axes::A
+end
```

Comment:
On the VectorizationBase side of things, it'd be nice if `StridedPointer` could be a `StrideLayout` + a `Ptr`.
However, `StrideLayout` is missing some things, most importantly the contiguous axis, and also has `axes` which it does not want (although it does want to store `map(static_first, axes(x))`).

Also, `StridedPointer` + `axes` == `PtrArray`. `StrideLayout` already subtypes `AbstractArray2`, so it is almost there itself.

Reply:
How about something that is indexable but isn't necessarily an iterator like this

```julia
struct StrideLayout{N,S<:Tuple{Vararg{Any,N}},R,O,O1,C}
    strides::S
    rank::R
    offset1::O1
    offsets::O
    contiguous::C
end
```

Reply:
I don't have any actual support anywhere for `ArrayInterface.contiguous_batch_size` yet, so we could add that later, but probably best to start carrying that around now as well.
But this looks good.